runners/jstorm/src/test/java/org/apache/beam/runners/jstorm/translation/PipelineTranslateTest.java

```diff
@@ -0,0 +1,184 @@
+    assertEquals(spoutName, stream.getProducer().getComponentId());
+    assertEquals(boltName, stream.getConsumer().getComponentId());
+    assertEquals(Stream.Grouping.Type.LOCAL_OR_SHUFFLE,
+        stream.getConsumer().getGrouping().getType());
+  }
+
+  @Test
+  public void testMultiParDoTranslation() {
+    JStormPipelineOptions options = PipelineOptionsFactory.as(JStormPipelineOptions.class);
+    Pipeline pipeline = createPipeline();
+    pipeline.apply(Create.of(1))
+        .apply("ParDo-1", ParDo.of(new TestDoFn()))
+        .apply("ParDo-2", ParDo.of(new TestDoFn()));
+
+    TranslationContext context = new TranslationContext(options);
+    JStormPipelineTranslator transformer = new JStormPipelineTranslator(context);
+    transformer.translate(pipeline);
+
+    // Source is translated to spout, and the tow ParDos are chained and translated to one bolt.
+    TranslationContext.ExecutionGraphContext executionGraphContext =
+        context.getExecutionGraphContext();
+    assertEquals(1, executionGraphContext.getSpouts().size());
+    assertEquals(1, executionGraphContext.getBolts().size());
+    assertEquals(1, Iterables.size(executionGraphContext.getStreams()));
+
+    // Check grouping type between spout and bolt
+    String spoutName = Iterables.getOnlyElement(executionGraphContext.getSpouts().keySet());
+    String boltName = Iterables.getOnlyElement(executionGraphContext.getBolts().keySet());
+    Stream stream = Iterables.getOnlyElement(executionGraphContext.getStreams());
+    assertEquals(spoutName, stream.getProducer().getComponentId());
+    assertEquals(boltName, stream.getConsumer().getComponentId());
+    assertEquals(Stream.Grouping.Type.LOCAL_OR_SHUFFLE,
+        stream.getConsumer().getGrouping().getType());
+  }
+
+  @Test
+  public void testGroupByKeyTranslation() {
+    JStormPipelineOptions options = PipelineOptionsFactory.as(JStormPipelineOptions.class);
+    Pipeline pipeline = createPipeline();
+    pipeline.apply("Source", Create.of(1))
+        .apply("ParDo", ParDo.of(new TestKvDoFn()))
+        .apply("GBK", GroupByKey.<String, Integer>create());
+
+    TranslationContext context = new TranslationContext(options);
+    JStormPipelineTranslator transformer = new JStormPipelineTranslator(context);
+    transformer.translate(pipeline);
+
+    // Source is translated to spout, ParDo and GBK are not chained and translated to two bolts.
+    TranslationContext.ExecutionGraphContext executionGraphContext =
+        context.getExecutionGraphContext();
+    assertEquals(1, executionGraphContext.getSpouts().size());
+    assertEquals(2, executionGraphContext.getBolts().size());
+    assertEquals(2, Iterables.size(executionGraphContext.getStreams()));
+
+    // Check grouping type between ParDo and GBK bolts
+    for (Stream stream: executionGraphContext.getStreams()) {
+      if (stream.getConsumer().getComponentId().indexOf("bolt") > -1
```

Comment:
Is this to figure out whether the consumers/producers are bolts or spouts?
This relies on the name convention, could we consider introducing getComponentType()?

Reply:
ExecutionGraph.getBolts can be used to check if is a spout or bolt.

Reply:
ExecutionGraph.getBolts can be used to check if is a spout or bolt.